### PR TITLE
Fixes #25756 - Adds Slot&Fill

### PR DESCRIPTION
**Pull request by amirfefer**
This PR implemented following to this [RFC](https://community.theforeman.org/t/rfc-expandable-component-architecture/12083)

- [x] Storybook Documentation

### CODE REVIEW

**Comment by amirfefer** [comment]
# Slot And Fill

`Slot & Fill` allows plugins to extend foreman core functionality in the UI

## Components

### Slot
`<Slot>` is an opinionated extension point which responsible on rendering its fills
a Slot can support multiple fills, rendering by weight order and can be toggled by the `multi` prop
if `multi` is off, it will render the maximum weighted fill.

### Fill
a fill is the filled object of a slot, can be a component or props object.

#### Component fill
a fill that contains a child  component, which rendered by a dedicated slot

*core*
```js
<Slot id='slot-id'>
 a default child // can be empty
</Slot>
```

*plugin A*
```js
<Fill id='slot-id' fillId='some-id' weight={100}> 
  <div> some text </ div>
</ Fill>
```

*plugin B*
```js
<Fill id='slot-id' fillId='some-id' weight={200}> 
  <div> some text </ div>
</ Fill>
```

Plugin B has a fill with a higher weight, therefore it will be rendered in a dedicated slot.

#### Component fill - remdering multiple fills
If a slot has a `multi` prop, and it has multiple fills, it will render these fills by weight order

*core*
```js
<Slot multi id='slot-id' />
```

*plugin A*
```js
<Fill id='slot-id' fillId='some-id' weight={100}> 
  <div> some text </ div>
</ Fill>
```

*plugin B*
```js
<Fill id='slot-id' fillId='some-id' weight={200}> 
  <div> some text </ div>
</ Fill>
```

Plugin B's fill will be render first

#### Props fill
a fill that contains an `overrideProps` object
those props are given to the slot's children


*core*
```js
const TextWrapper = ({ text }) => <div>{text}</div>; 

<Slot id='slot-id'>
  <TextWrapper text='some default text' />
</Slot>
```

*plugin A*
```js
<Fill id='slot-id' fillId='some-id' weight={200} overrideProps={{ text: '[Plugin A] this text given by a prop' }} /> 
```


*plugin B*
```js
<Fill id='slot-id' fillId='some-id' weight={100} overrideProps={{ text: '[Plugin B] this text given by a prop' }} /> 
```
In this case, the slot doesn't have `multi` prop, therefore it will take the max weight which is Plugin A's fill.

#### Global fill
This fill is available on each and every core page

In order to create a global fill, first create `fills_index.js` file under webpack directory:
```js
import React from 'react';
import SomeComponent from './components/SomeComponent';
import { AddGlobalFill } from 'foremanReact/components/common/Fill/GlobalFill';

AddGlobalFill('slotId', 'fillId', <SomeComponent key="1" />, 300);
```
*core*
Add a slot in foreman core page by importing `Slot` or by using `slot` helper in erb file:
```ruby
  <%= slot('slotId', true) %>
```

File: webpack/assets/javascripts/react_app/components/common/Fill/Fill.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  }
+}
+
+Fill.propTypes = {
+  children: PropTypes.oneOfType([PropTypes.node, PropTypes.object]),
+  registerFillComponent: PropTypes.func.isRequired,
+  unregisterFillComponent: PropTypes.func.isRequired,
+  id: PropTypes.string.isRequired,
+  weight: PropTypes.number.isRequired,
+  fillId: PropTypes.string.isRequired,
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/Fill.js]: Maybe it's a nitpick but it feels to me like the `id` prop should represent the `filleId` since this is the `Fill` component.
The current `fillId` can be changed to `slotId` and then it would make more sense.

File: webpack/assets/javascripts/react_app/components/common/Fill/FillConstants.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export const REGISTER_FILL = '[SLOT & FILL] REGISTER_FILL';
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/FillConstants.js]: What is the reason not to use our current convention, e.g. `SLOT_AND_FILL_REGISTER_FILL`?

File: webpack/assets/javascripts/react_app/components/common/Fill/GlobalFill.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { registerFillComponent } from '../Fill/FillActions';
+import store from '../../../redux';
+
+export const AddGlobalFill = (slotId, fillId, component, weight) => {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/GlobalFill.js]: nitpick: `addGlobalFill`

File: webpack/assets/javascripts/react_app/components/common/Fill/__tests__/FillActions.test.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { testActionSnapshotWithFixtures } from '../../../../common/testHelpers';
+import { registerFillComponent, unregisterFillComponent } from '../FillActions';
+
+const conponent = <div> a component </div>;
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/__tests__/FillActions.test.js]: Notice it's not a component, shouldn't it suppose to be:
```jsx
const SomeComponent = () => <div>a component</div>;
```

File: webpack/assets/javascripts/react_app/components/common/Fill/index.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import { connect } from 'react-redux';
+
+import * as actions from './FillActions';
+import reducer from './FillReducer';
+
+import Fill from './Fill';
+
+// map state to props
+const mapStateToProps = state => ({
+  components: state.extendable,
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/index.js]: Don't we want to use a selector so we can reuse the statment `state.extendable`?

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  };
+
+  if (Object.keys(fills).length) {
+    if (multi) {
+      return selectComponentByWeight(id).map(component => addProps(component));
+    }
+    return addProps(selectMaxComponent(id));
+  }
+  if (children) return children;
+  return null;
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.js]: Can be `return children || null`.
Or `children` default value can be set to `null`

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        </Fill>
+
+        <Slot id="slot-1" multi>
+          <div> defaul value </div>
+        </Slot>
+      </React.Fragment>
+    );
+
+    wrapper.update();
+    expect(toJson(wrapper.find('Slot'))).toMatchSnapshot();
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js:34]: Don't we want to take a snapshot of the `state` and the called `actions` to make sure the whole lifecycle worked well?

File: webpack/assets/javascripts/react_app/components/common/Slot/SlotSelectors.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { orderBy } from 'lodash';
+import { getComponentsBySlotId } from '../../../../services/ExtendableRegistry';
+
+export const selectComponentByWeight = slotId =>
+  orderBy(getComponentsBySlotId(slotId), ['weight'], ['desc']).map(
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Slot/SlotSelectors.js]: Could you use [sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) instead of using `lodash`?

File: webpack/assets/javascripts/react_app/components/componentRegistry.js [hunk 9]
```diff
@@ -123,6 +124,7 @@ const coreComponets = [
   { name: 'ConfigReports', type: ConfigReports },
   { name: 'DiffModal', type: DiffModal },
   { name: 'TemplateInput', type: TemplateInput },
+  { name: 'Slot', type: Slot },
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/componentRegistry.js:129]: What about `Fill`?

File: webpack/assets/javascripts/services/ExtendableRegistry/index.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+const components = {};
```
Comment (sharvit) [inline webpack/assets/javascripts/services/ExtendableRegistry/index.js]: About this file:

The name is a bit confusing if we have `Slot` and `Fill` what is `ExtendableRegistry`?
What about calling this file `SlotsRegistry`?

About `add` and `remove`, we can name them as `fillSlot` and `unfillSlot`?

File: webpack/assets/javascripts/services/ExtendableRegistry/index.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  components[SlotId][fillId] = { component, weight };
+};
+
+export const remove = (SlotId, fillId) => {
+  const slotItems = components[SlotId];
+
+  delete slotItems[fillId];
+};
+
+export const getComponentsBySlotId = id => Object.values(components[id]);
```
Comment (sharvit) [inline webpack/assets/javascripts/services/ExtendableRegistry/index.js]: What about naming it `getSlotComponents`?

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @amirfefer, Awesome work :+1:

File: webpack/assets/javascripts/react_app/components/common/Fill/Fill.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    const { id, unregisterFillComponent, fillId } = this.props;
+
+    unregisterFillComponent(id, fillId);
+  }
+  render() {
+    return null;
+  }
+}
+
+Fill.propTypes = {
```
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/common/Fill/Fill.js:27]: Could we briefly inline document the props?

File: webpack/stories/docs/SlotAndFill.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+*plugin B*
+```js
+<Fill id='slot-id' fillId='some-id' weight={100} overrideProps={{ text: '[Plugin B] this text given by a prop' }} /> 
+```
+In this case, the slot doesn't have `multi` prop, therefore it will take the max weight which is Plugin A's fill.
+
+#### Global fill
+This fill is available on each and every core page
+
+create a `fills_index.js` file under webpack directory:
```
Comment (xprazak2) [inline webpack/stories/docs/SlotAndFill.md:96]: fills_index is not picked up by default, it needs to be imported into webpack/index.js

```js
import './fills_index';
```

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
+import { cloneElement, isValidElement } from 'react';
+import PropTypes from 'prop-types';
+import { selectMaxComponent, selectComponentByWeight } from './SlotSelectors';
+
+const Slot = ({ fills, id, multi, children, ...props }) => {
+  const addProps = object => {
+    if (multi && !object.key) {
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.js:6]: Is `key` needed only because `Each child in a list should have a unique "key" prop` or do we use it somewhere directly?

Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.js:6]: I believe the main thing is the `each child` thing, but that is fairly important as without it React is unable to do its mojo and is loosing a lot of performance tweeks. So I would agree there should be an warning about that - the react one doesn't need to kick in as I can develop my plugin in isolation.
Maybe console warning would be sufficient though?

File: webpack/stories/docs/SlotAndFill.md (hunk 13 above)
Comment (sharvit) [reply webpack/stories/docs/SlotAndFill.md:96]: I think `fills_index` will be compiled by `webpack` into its own bundle since it ends with `*_index.js`.
Then it should be loaded into core layout in some way?
@amirfefer more information needed in this md file?

File: webpack/assets/javascripts/react_app/components/common/Fill/FillConstants.js (hunk 2 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Fill/FillConstants.js]: I find this a bit more readable when debugging, but in this PR i won't change any convention :+1:

File: webpack/assets/javascripts/react_app/components/componentRegistry.js (hunk 9 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/componentRegistry.js:129]: Fill is a registration component, therefore it doesn't make sense to add it on erb  (a consumer can use globalFill instead, or create a component with a fill and a desire component to register)

File: webpack/assets/javascripts/services/ExtendableRegistry/index.js (hunk 10 above)
Comment (amirfefer) [reply webpack/assets/javascripts/services/ExtendableRegistry/index.js]: I renamed to `add` and `remove` due to a prior review, I think it clear enough , plus this implementation is transparent to the consumers

File: webpack/stories/docs/SlotAndFill.md (hunk 13 above)
Comment (amirfefer) [reply webpack/stories/docs/SlotAndFill.md:96]: fills_index is picked  by webpack automatically https://github.com/theforeman/foreman/pull/6554
But a registration is still needed

```ruby 
# plugin.rb
register_global_file 'fills'
```

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js (hunk 14 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.js:6]: Yes, the key is needed due to react architecture, and I agree about console warning in this case :+1:

**Comment by amirfefer** [comment]
Thanks @sharvit, @xprazak2  and @ezr-ondrej ! 

ready for another look :)

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js (hunk 7 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js:34]: Absolutely ! I added a full redux's lifecycle test

File: webpack/stories/docs/SlotAndFill.md (hunk 13 above)
Comment (amirfefer) [reply webpack/stories/docs/SlotAndFill.md:96]: I added this information to the md file :+1:

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import { cloneElement, isValidElement } from 'react';
+import PropTypes from 'prop-types';
+import { selectMaxComponent, selectComponentByWeight } from './SlotSelectors';
+
+const Slot = ({ fills, id, multi, children = null, ...props }) => {
+  const addProps = object => {
+    if (multi && !object.key) {
+      // eslint-disable-next-line no-console
+      console.warnn('Please add a key attriubte to multiple fills');
```
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.js]: Probably a typo? :)

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js (hunk 14 above)
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.js:6]: Would it make sense to make `Slot` responsible for creating the keys somehow to make sure they are unique? My thinking is that a developer needs to know about all the places that use a particular slot to make sure the value for a key is really unique, which is really difficult when I consider just the number of 'official' plugins...

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js (hunk 15 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.js]: good catch ! 
Fixed :)

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js (hunk 14 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.js:6]: I think that a component should be responsible for its key (I prefer a domain related key than a random key), if the key isn't unique, then a warning will appear at the console, so during the developing or reviewing it would be easy to find out

**Comment by amirfefer** [comment]
The first PR that uses slot&fill - https://github.com/Katello/katello/pull/8039 :)

**Comment by amirfefer** [comment]
tests failure seem unrelated

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js (hunk 14 above)
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.js:6]: I guess, I was just wondering if we could make it as easy as possible to use, but maybe adding a note to the docs explaining the key should be unique and there will be warning if it is not will be enough.

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
+import { cloneElement, isValidElement } from 'react';
+import PropTypes from 'prop-types';
+import { selectMaxComponent, selectComponentByWeight } from './SlotSelectors';
+
+const Slot = ({ fills, id, multi, children = null, ...props }) => {
+  const addProps = object => {
+    if (multi && !object.key) {
+      // eslint-disable-next-line no-console
+      console.warn('Please add a key attriubte to multiple fills');
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.js]: nit: s/attriubte/attribute/

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+        <React.Fragment>
+          <Fill slotId="slot-7" id="some-key-1" weight={100} />
+
+          <Slot id="slot-7" />
+        </React.Fragment>
+      );
+    }).toThrowError(new Error('Slot with override props must have a child'));
+  });
+
+  it('no nultiple fills', () => {
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js]: nit: s/nultiple/multiple/

**Comment by xprazak2** [comment]
Nice progress, I added a couple of comments.

**Comment by xprazak2** [comment]
Awesome progress, just a few minor things.

File: app/views/about/index.html.erb [hunk 18]
```diff
@@ -140,5 +140,6 @@
       <p id="copyright-p"><%= (_("Version %{version}  %{copyright} 2009-%{year} Paul Kelly and %{author}") % {:version => SETTINGS[:version], :copyright => "&copy;", :year=>DateTime.now.year, :author=>mail_to("ohadlevy@gmail.com", "Ohad Levy" )}).html_safe %></p>
     </div>
   </div>
+  <%= slot('aboutFooterSlot', true) %>
 </div>
-</div>
```
Comment (sharvit) [inline app/views/about/index.html.erb:144]: Why did you remove the closing div tag? Is this aim to fix something different than the scope of this PR? Or did it happen by accident?

File: webpack/assets/javascripts/react_app/components/common/Fill/FillReducer.js [hunk 19]
```diff
@@ -0,0 +1,27 @@
+import Immutable from 'seamless-immutable';
+
+import { REGISTER_FILL, REMOVE_FILLED_COMPONENT } from './FillConstants';
+
+const initialState = Immutable({
+  RegisteredComponent: {},
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/FillReducer.js]: IMO `RegisteredComponent` is redundant, it could be:
```jsx
const initialState = Immutable({});

return state.setIn([payload.slotId, payload.fillId], true);
```

File: webpack/assets/javascripts/react_app/components/common/Fill/GlobalFill.js [hunk 20]
```diff
@@ -0,0 +1,8 @@
+import { registerFillComponent } from '../Fill/FillActions';
+import store from '../../../redux';
+
+export const addGlobalFill = (slotId, fillId, component, weight) => {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/GlobalFill.js:4]: Don't you want to allow `globalFills` with `overrideProps`?

File: webpack/assets/javascripts/react_app/components/common/Fill/index.js [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+import { selectExtendableComponents } from './FillSelectors';
+
+import * as actions from './FillActions';
+import reducer from './FillReducer';
+
+import Fill from './Fill';
+
+// map state to props
+const mapStateToProps = state => ({
+  components: selectExtendableComponents(state),
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/index.js]: I'm confused, couldn't find any `components` prop the `Fill` using.

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    if (!children) {
+      throw new Error('Slot with override props must have a child');
+    }
+
+    return cloneElement(children, { ...props, ...object });
+  };
+
+  if (Object.keys(fills).length) {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.js]: Can we extract this logic to the "selectors" so this component will receive the `end-result` of those selectors as a prop?

File: webpack/assets/javascripts/services/ExtendableRegistry/index.js [hunk 23]
```diff
@@ -0,0 +1,17 @@
+const slotsRegistry = {};
```
Comment (sharvit) [inline webpack/assets/javascripts/services/ExtendableRegistry/index.js]: Could you rename the filename to `slotsRegistry` so developers would know it's apart of the slot&fill?

File: webpack/stories/docs/SlotAndFill.md [hunk 24]
```diff
@@ -0,0 +1,116 @@
+# Slot And Fill
+
+`Slot & Fill` allows plugins to extend foreman core functionality in the UI
+
+## Components
+
+### Slot
+`<Slot>` is an opinionated extension point which responsible of rendering its fills
+a Slot can support multiple fills, rendering by weight order by the `multi` prop
+if there is no such prop, it will render the max weighted dfill.
```
Comment (sharvit) [inline webpack/stories/docs/SlotAndFill.md]: dfill should be fill?

File: webpack/stories/docs/SlotAndFill.md [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+If a slot has a `multi` prop, and it has multiple fills, it will render these fills by weight order
+
+*core*
+```js
+<Slot multi id='slot-id' />
+```
+
+*plugin A*
+```js
+<Fill id='slot-id' fillId='some-id' weight={100}>
```
Comment (sharvit) [inline webpack/stories/docs/SlotAndFill.md]: Didn't you change it so there are `id` and `slotId` now?

File: webpack/stories/docs/SlotAndFill.md [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+const TextWrapper = ({ text }) => <div>{text}</div>; 
+
+<Slot id='slot-id'>
+  <TextWrapper text='some default text' />
+</Slot>
+```
+
+*plugin A*
+```js
+<Fill id='slot-id' fillId='some-id' weight={200} overrideProps={{ text: '[Plugin A] this text given by a prop' }} />
```
Comment (sharvit) [inline webpack/stories/docs/SlotAndFill.md]: Can I use the `overrideProps` in a way it will extend the original prop?
For example, what if I want to add a prefix to the text? so it will end up as `bla bla some default text`?

File: webpack/stories/docs/SlotAndFill.md [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+
+
+*plugin B*
+```js
+<Fill id='slot-id' fillId='some-id' weight={100} overrideProps={{ text: '[Plugin B] this text given by a prop' }} /> 
+```
+In this case, the slot doesn't have `multi` prop, therefore it will take the max weight which is Plugin A's fill.
+
+#### Global fill
+This fill is available on each and every core page
```
Comment (sharvit) [inline webpack/stories/docs/SlotAndFill.md:92]: Can you expand about why this is needed and what problems it is trying to solve?
I want developers to know when they should and shouldn't use it.

File: webpack/stories/docs/SlotAndFill.md [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+
+create a `fills_index.js` file under webpack directory:
+```js
+import React from 'react';
+import SomeComponent from './components/SomeComponent';
+import { AddGlobalFill } from 'foremanReact/components/common/Fill/GlobalFill';
```
Comment (sharvit) [inline webpack/stories/docs/SlotAndFill.md]: nitpick: `addGlobalFill` (you already changed it in the original file)

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @amirfefer :+1:

File: app/views/about/index.html.erb (hunk 18 above)
Comment (amirfefer) [reply app/views/about/index.html.erb:144]: this is an extra closing div (with no starting one)

File: webpack/assets/javascripts/react_app/components/common/Fill/GlobalFill.js (hunk 20 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Fill/GlobalFill.js:4]: see at the documentation, the component can be a also a props object

File: webpack/stories/docs/SlotAndFill.md (hunk 26 above)
Comment (amirfefer) [reply webpack/stories/docs/SlotAndFill.md]: overrideProps by its name will override the slot children props, so basically it replaces the content

File: webpack/stories/docs/SlotAndFill.md (hunk 27 above)
Comment (amirfefer) [reply webpack/stories/docs/SlotAndFill.md:92]: sure ,I'll add a sentence or two about it

File: webpack/stories/docs/SlotAndFill.md (hunk 26 above)
Comment (sharvit) [reply webpack/stories/docs/SlotAndFill.md]: So how would I implement what I have asked before?
I think this type of example is necessary for the documentation.

Comment (amirfefer) [reply webpack/stories/docs/SlotAndFill.md]: sure :+1: , Added in the documentation

File: webpack/assets/javascripts/react_app/components/common/Fill/index.js (hunk 21 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Fill/index.js]: doesn't needed anymore, thanks !

**Comment by amirfefer** [comment]
Thanks @xprazak2 and @sharvit :+1: 
Updated

**Comment by amirfefer** [comment]
rebased

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js [hunk 29]
```diff
@@ -0,0 +1,185 @@
...
+        </Slot>
+      </React.Fragment>
+    );
+    wrapper.update();
+    expect(toJson(wrapper.find('Slot'))).toMatchSnapshot();
+  });
+
+  it('check redux lifecycle when rendering fills on a slot', () => {
+    integrationTestHelper.takeStoreSnapshot();
+    integrationTestHelper.takeActionsSnapshot();
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js]: What benefit do those snapshots give?
I think the point is to use them inside the other tests so together with your `jsx` snapshots you have a snapshot of the action and the state that cause your `jsx` snapshot.

File: webpack/assets/javascripts/services/ExtendableRegistry/index.js (hunk 23 above)
Comment (sharvit) [reply webpack/assets/javascripts/services/ExtendableRegistry/index.js]: ?

File: webpack/assets/javascripts/react_app/components/common/Fill/FillReducer.js (hunk 19 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/common/Fill/FillReducer.js]: ?

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @amirfefer

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js (hunk 29 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js]: Look at the snapshot that this has created, basically it got all of the actions and corresponding reducers, so with this we test the entire functionality of redux

Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/common/Slot/Slot.test.js]: Yes but it creates snapshots in this test but they based on the logic ran in the other tests :/
Therefore I believe each test should use his own `IntegrationTestHelper` instance and keep its own state snapshots (that related to the context of the test).

**Review by sharvit — APPROVED** [review]
Looks good and work as expected!
Tested with https://github.com/Katello/katello/pull/8039

Thanks @amirfefer :postal_horn:

**Comment by xprazak2** [comment]
I saw the snapshot failures on other PRs as well, so probably not related. Seems like all comments were addressed. Any additional concerns?

**Comment by sharvit** [comment]
> I saw the snapshot failures on other PRs as well, so probably not related. Seems like all comments were addressed. Any additional concerns?

I can think about any and I feel like it is ready :+1:

**Comment by sharvit** [comment]
> > I saw the snapshot failures on other PRs as well, so probably not related. Seems like all comments were addressed. Any additional concerns?
> 
> I can think about any and I feel like it is ready

**Can't**

**Review by xprazak2 — APPROVED** [review]

**Comment by xprazak2** [comment]
Let's get this in. Thanks everyone!
